// @salesforce/aura-language-server/lib/markup/auraTags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAuraTagProvider = exports.setIndexer = void 0;
let indexer;
function getAuraTags() {
    if (indexer) {
        return indexer.getAuraTags();
    }
    return new Map();
}
function getAuraByTag(tag) {
    if (indexer) {
        return indexer.getAuraByTag(tag);
    }
    return undefined;
}
function setIndexer(idx) {
    indexer = idx;
}
exports.setIndexer = setIndexer;
function getTagsData() {
    return Array.from(getAuraTags()).map(([tag, tagInfo]) => ({
        name: tag,
        description: tagInfo.getHover(),
        attributes: tagInfo.attributes.map((attr) => ({
            name: attr.name,
            description: attr.name,
            valueSet: attr.type,
        })),
    }));
}
function getAttributesData(tag) {
    const cTag = getAuraByTag(tag);
    if (cTag) {
        return cTag.attributes.map((attr) => ({
            name: attr.name,
            description: attr.name,
            valueSet: attr.type,
        }));
    }
    return [];
}
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function getValuesData(tag, attribute) {
    // TODO provide suggestions by consulting shapeService
    return [];
}
function getAuraTagProvider() {
    return {
        getId: () => 'aura',
        isApplicable: (languageId) => languageId === 'html',
        provideTags: getTagsData,
        provideAttributes: getAttributesData,
        provideValues: getValuesData,
    };
}
exports.getAuraTagProvider = getAuraTagProvider;
