// @salesforce/aura-language-server/lib/aura-server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const vscode_languageserver_1 = require("vscode-languageserver");
const vscode_html_languageservice_1 = require("vscode-html-languageservice");
const vscode_uri_1 = __importDefault(require("vscode-uri"));
const lightning_lsp_common_1 = require("@salesforce/lightning-lsp-common");
const tern_server_1 = require("./tern-server/tern-server");
const indexer_1 = __importDefault(require("./aura-indexer/indexer"));
const utils_1 = require("@salesforce/lightning-lsp-common/lib/utils");
const auraTags_1 = require("./markup/auraTags");
const shared_1 = require("@salesforce/lightning-lsp-common/lib/shared");
const aura_utils_1 = require("./aura-utils");
const tagAdded = new vscode_languageserver_1.NotificationType('salesforce/tagAdded');
const tagDeleted = new vscode_languageserver_1.NotificationType('salesforce/tagDeleted');
const tagsCleared = new vscode_languageserver_1.NotificationType('salesforce/tagsCleared');
class Server {
    connection = (0, vscode_languageserver_1.createConnection)();
    documents = new vscode_languageserver_1.TextDocuments();
    context;
    workspaceFolders;
    workspaceRoots;
    htmlLS;
    auraIndexer;
    constructor() {
        (0, lightning_lsp_common_1.interceptConsoleLogger)(this.connection);
        this.connection.onInitialize(this.onInitialize.bind(this));
        this.connection.onCompletion(this.onCompletion.bind(this));
        this.connection.onCompletionResolve(this.onCompletionResolve.bind(this));
        this.connection.onHover(this.onHover.bind(this));
        this.connection.onDefinition(this.onDefinition.bind(this));
        this.connection.onTypeDefinition(this.onTypeDefinition.bind(this));
        this.connection.onDidChangeWatchedFiles(this.onDidChangeWatchedFiles.bind(this));
        this.connection.onRequest('salesforce/listComponents', this.onListComponents.bind(this));
        this.connection.onRequest('salesforce/listNamespaces', this.onListNamespaces.bind(this));
        this.documents.listen(this.connection);
        this.documents.onDidClose(this.onDidClose.bind(this));
        this.documents.onDidOpen(tern_server_1.addFile);
        this.documents.onDidChangeContent(tern_server_1.addFile);
        this.documents.onDidClose(tern_server_1.delFile);
        this.connection.onReferences(tern_server_1.onReferences);
        this.connection.onSignatureHelp(tern_server_1.onSignatureHelp);
    }
    async onInitialize(params) {
        const { workspaceFolders } = params;
        this.workspaceFolders = workspaceFolders;
        this.workspaceRoots = workspaceFolders.map((folder) => path.resolve(vscode_uri_1.default.parse(folder.uri).fsPath));
        try {
            if (this.workspaceRoots.length === 0) {
                console.warn('No workspace found');
                return { capabilities: {} };
            }
            for (const root of this.workspaceRoots) {
                console.info(`Starting *AURA* language server at ${root}`);
            }
            const startTime = process.hrtime();
            this.context = new lightning_lsp_common_1.WorkspaceContext(this.workspaceRoots);
            if (this.context.type === shared_1.WorkspaceType.CORE_PARTIAL) {
                await (0, tern_server_1.startServer)(path.join(this.workspaceRoots[0], '..'), path.join(this.workspaceRoots[0], '..'));
            }
            else {
                await (0, tern_server_1.startServer)(this.workspaceRoots[0], this.workspaceRoots[0]);
            }
            this.context.configureProject();
            this.auraIndexer = new indexer_1.default(this.context);
            (0, auraTags_1.setIndexer)(this.auraIndexer);
            this.setupIndexerEvents();
            this.startIndexing();
            this.htmlLS = (0, vscode_html_languageservice_1.getLanguageService)();
            this.htmlLS.setDataProviders(true, [(0, auraTags_1.getAuraTagProvider)()]);
            console.info('... language server started in ' + lightning_lsp_common_1.utils.elapsedMillis(startTime));
            return {
                capabilities: {
                    textDocumentSync: this.documents.syncKind,
                    completionProvider: {
                        resolveProvider: true,
                        triggerCharacters: ['.', ':', '<', '"', '=', '/', '>'],
                    },
                    workspace: {
                        workspaceFolders: {
                            supported: true,
                        },
                    },
                    signatureHelpProvider: {
                        triggerCharacters: ['('],
                    },
                    referencesProvider: true,
                    hoverProvider: true,
                    definitionProvider: true,
                    typeDefinitionProvider: true,
                },
            };
        }
        catch (e) {
            throw new Error(`Aura Language Server initialization unsuccessful. Error message: ${e.message}`);
        }
    }
    setupIndexerEvents() {
        this.auraIndexer.eventEmitter.on('set', (tag) => {
            this.connection.sendNotification(tagAdded, { taginfo: tag });
        });
        this.auraIndexer.eventEmitter.on('delete', (tag) => {
            this.connection.sendNotification(tagDeleted, tag);
        });
        this.auraIndexer.eventEmitter.on('clear', () => {
            this.connection.sendNotification(tagsCleared, undefined);
        });
    }
    startIndexing() {
        setTimeout(async () => {
            this.connection.sendNotification('salesforce/indexingStarted');
            await this.auraIndexer.configureAndIndex();
            this.connection.sendNotification('salesforce/indexingEnded');
        }, 0);
    }
    async onCompletion(completionParams) {
        const document = this.documents.get(completionParams.textDocument.uri);
        if (await this.context.isAuraMarkup(document)) {
            const htmlDocument = this.htmlLS.parseHTMLDocument(document);
            const list = this.htmlLS.doComplete(document, completionParams.position, htmlDocument, {
                isSfdxProject: this.context.type === shared_1.WorkspaceType.SFDX,
                useAttributeValueQuotes: true,
            });
            return list;
        }
        if (await this.context.isAuraJavascript(document)) {
            return (0, tern_server_1.onCompletion)(completionParams);
        }
        return { isIncomplete: false, items: [] };
    }
    onCompletionResolve(item) {
        return item;
    }
    async onHover(textDocumentPosition) {
        const document = this.documents.get(textDocumentPosition.textDocument.uri);
        if (await this.context.isAuraMarkup(document)) {
            const htmlDocument = this.htmlLS.parseHTMLDocument(document);
            const hover = this.htmlLS.doHover(document, textDocumentPosition.position, htmlDocument);
            return hover;
        }
        if (await this.context.isAuraJavascript(document)) {
            return (0, tern_server_1.onHover)(textDocumentPosition);
        }
        return null;
    }
    async onTypeDefinition(textDocumentPosition) {
        const document = this.documents.get(textDocumentPosition.textDocument.uri);
        if (await this.context.isAuraJavascript(document)) {
            return (0, tern_server_1.onTypeDefinition)(textDocumentPosition);
        }
        return null;
    }
    findJavascriptProperty(valueProperty, textDocumentPosition) {
        // couldn't find it within the markup file, try looking for it as a javascript property
        const fsPath = vscode_uri_1.default.parse(textDocumentPosition.textDocument.uri).fsPath;
        const parsedPath = path.parse(fsPath);
        const componentName = parsedPath.name;
        const namespace = path.basename(path.dirname(parsedPath.dir));
        const tag = this.auraIndexer.getAuraByTag(namespace + ':' + componentName);
        if (tag) {
            // aura tag doesn't contain controller methods yet
            // but, if its not a v.value, its probably fine to just open the controller file
            const controllerPath = path.join(parsedPath.dir, componentName + 'Controller.js');
            return {
                uri: vscode_uri_1.default.file(controllerPath).toString(),
                range: {
                    start: {
                        character: 0,
                        line: 1,
                    },
                    end: {
                        character: 0,
                        line: 1,
                    },
                },
            };
        }
        return null;
    }
    async onDefinition(textDocumentPosition) {
        const document = this.documents.get(textDocumentPosition.textDocument.uri);
        if (await this.context.isAuraMarkup(document)) {
            const htmlDocument = this.htmlLS.parseHTMLDocument(document);
            const def = (0, aura_utils_1.getAuraBindingTemplateDeclaration)(document, textDocumentPosition.position, htmlDocument);
            if (def) {
                return def;
            }
            const valueProperty = (0, aura_utils_1.getAuraBindingValue)(document, textDocumentPosition.position, htmlDocument);
            if (valueProperty) {
                return this.findJavascriptProperty(valueProperty, textDocumentPosition);
            }
        }
        if (await this.context.isAuraJavascript(document)) {
            return (0, tern_server_1.onDefinition)(textDocumentPosition);
        }
        return null;
    }
    async onDidChangeWatchedFiles(change) {
        console.info('aura onDidChangeWatchedFiles...');
        const changes = change.changes;
        try {
            if (lightning_lsp_common_1.utils.isAuraRootDirectoryCreated(this.context, changes)) {
                await this.context.getIndexingProvider('aura').resetIndex();
                await this.context.getIndexingProvider('aura').configureAndIndex();
                // re-index everything on directory deletions as no events are reported for contents of deleted directories
                const startTime = process.hrtime();
                console.info('reindexed workspace in ' + lightning_lsp_common_1.utils.elapsedMillis(startTime) + ', directory was deleted:', changes);
                return;
            }
            else {
                for (const event of changes) {
                    if (event.type === vscode_languageserver_1.FileChangeType.Deleted && lightning_lsp_common_1.utils.isAuraWatchedDirectory(this.context, event.uri)) {
                        const dir = (0, utils_1.toResolvedPath)(event.uri);
                        this.auraIndexer.clearTagsforDirectory(dir, this.context.type === shared_1.WorkspaceType.SFDX);
                    }
                    else {
                        const file = (0, utils_1.toResolvedPath)(event.uri);
                        if (file.endsWith('.app') || file.endsWith('.cmp') || file.endsWith('.intf') || file.endsWith('.evt') || file.endsWith('.lib')) {
                            await this.auraIndexer.indexFile(file, this.context.type === shared_1.WorkspaceType.SFDX);
                        }
                    }
                }
            }
        }
        catch (e) {
            this.connection.sendNotification(vscode_languageserver_1.ShowMessageNotification.type, {
                type: vscode_languageserver_1.MessageType.Error,
                message: `Error re-indexing workspace: ${e.message}`,
            });
        }
    }
    onListComponents() {
        const tags = this.auraIndexer.getAuraTags();
        const result = JSON.stringify([...tags]);
        return result;
    }
    onListNamespaces() {
        const tags = this.auraIndexer.getAuraNamespaces();
        const result = JSON.stringify(tags);
        return result;
    }
    onDidClose(event) {
        this.connection.sendDiagnostics({ uri: event.document.uri, diagnostics: [] });
    }
    listen() {
        this.connection.listen();
    }
}
exports.default = Server;
