// @lwc/metadata/dist/script-metadata/visitors/module-imports-visitor.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var reference = require('../module-references/reference.js');
var utils$1 = require('../../shared/utils.js');
var utils = require('../utils.js');
var resolver = require('../resolver.js');

// Copyright (c) 2022, Salesforce, Inc.,
// All rights reserved.
// For full license text, see the LICENSE.txt file
function gatherImportMetadata(node, result, refId) {
    const { source: { value: moduleSpecifier }, specifiers, importKind, } = node;
    const importMetadata = {
        moduleSpecifier,
        location: utils.generateSourceLocation(node),
        refId,
    };
    // add import kind for type imports in TypeScript files
    if (importKind) {
        importMetadata.importKind = importKind;
    }
    const namedImports = [];
    specifiers.forEach((specifier) => {
        const { local } = specifier;
        const location = utils.generateSourceLocation(specifier);
        switch (specifier.type) {
            case 'ImportDefaultSpecifier': {
                importMetadata.defaultBinding = {
                    name: local.name,
                    location,
                };
                break;
            }
            case 'ImportNamespaceSpecifier': {
                importMetadata.namespaceImport = {
                    aliasName: local.name,
                    location,
                };
                break;
            }
            case 'ImportSpecifier': {
                const { imported, importKind } = specifier;
                const importedName = resolver.resolveImportSpecifierImported(imported);
                const namedImport = {
                    name: importedName,
                    aliasName: local.name === importedName ? undefined : local.name,
                    location,
                };
                if (importKind) {
                    namedImport.importKind = importKind;
                }
                namedImports.push(namedImport);
                break;
            }
        }
    });
    if (!utils$1.isEmpty(namedImports)) {
        importMetadata.namedImports = namedImports;
    }
    result.imports.push(importMetadata);
}
function moduleImportVisitor(state, result) {
    return function ImportVisitorFunction(path) {
        const { node } = path;
        const moduleReference = reference.gatherModuleReference(node, state, result);
        gatherImportMetadata(node, result, moduleReference.id);
    };
}

exports.default = moduleImportVisitor;
